acme/acme.go

```diff
@@ -328,29 +355,44 @@ func (a *ACME) CreateLocalConfig(tlsConfig *tls.Config, checkOnDemandDomain func
 func (a *ACME) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
 	domain := types.CanonicalDomain(clientHello.ServerName)
 	account := a.store.Get().(*Account)
-	//use regex to test for wildcard certs that might have been added into TLSConfig
-	for k := range a.TLSConfig.NameToCertificate {
-		selector := "^" + strings.Replace(k, "*.", "[^\\.]*\\.?", -1) + "$"
-		match, _ := regexp.MatchString(selector, domain)
-		if match {
-			return a.TLSConfig.NameToCertificate[k], nil
-		}
-	}
+
+	// Check for existing challenge cert
+	log.Debugf("Checking ACME challenge for %s", domain)
 	if challengeCert, ok := a.challengeProvider.getCertificate(domain); ok {
-		log.Debugf("ACME got challenge %s", domain)
+		log.Debugf("Returning ACME challenge cert for %s", domain)
 		return challengeCert, nil
 	}
+
+	// Check for existing ACME domain cert
+	log.Debugf("Checking ACME domain cert for %v", domain)
 	if domainCert, ok := account.DomainsCertificate.getCertificateForDomain(domain); ok {
-		log.Debugf("ACME got domain cert %s", domain)
+		log.Debugf("Returning ACME domain cert for %s", domain)
 		return domainCert.tlsCert, nil
 	}
+
 	if a.OnDemand {
 		if a.checkOnDemandDomain != nil && !a.checkOnDemandDomain(domain) {
 			return nil, nil
 		}
-		return a.loadCertificateOnDemand(clientHello)
+		cert, err := a.loadCertificateOnDemand(clientHello)
+		if cert != nil && err == nil {
+			return cert, nil
+		}
+		log.Errorf("On demand certicate retrival failed for %v due to %v", domain, err)
 	}
-	log.Debugf("ACME got nothing %s", domain)
+
+	// Check if we have a wildcard cert into TLSConfig that could be used
+	log.Debugf("Checking wildcard certificate matching for %v", domain)
+	for k := range a.TLSConfig.NameToCertificate {
```

Comment:
No problem :)

If I understand correctly the modifications you did, the `getCertificate` method will follow this new workflow :

1. Check the existence of a ACME challenge certificate for the current domain, if the certicate exists,returns otherwise,
2. Check the existence of a ACME domain certificate for the current domain, if the certicate exists,returns otherwise,
3. If the `OnDemand` parameter is true AND if the domain route must be checked on demand AND if the domain does not respect the `IgnoreFilter` regex, generate a new ACME certificate and return it otherwise,
4. Check the existence of a wildcard certificate which can validate the current domain, if the certicate exists,returns otherwise,
5. Return none certificate.

With this algorithm, all users who have a wildcard certificate and who want to use it for all the (sub)domains, have to specify a regex in the `IgnoreFilter` to match with the wildcard certificate.
This is a new behaviour, with a new parameter, and users who won't be aware with it will generate an ACME certificate for all their domains if the `OnDemand` option is activated => It breaks the backward compatibility.

With the suggestion I did, only users who activate the new option `IgnoreFilter` would follow this new worflow, the others will continue with the current one. But this solution was nor perfect.
That's why I discussed about this it with @ldez, @juliens and @emile.

**Finally, we would suggest you to use the `IgnoreFilter` in the way to apply it to the domains which not have to be checked by a wildcard certificate.**
With this modification, the workflow in the `generateCertificate` method will be the same, but users will be able to specify (sub)domains for which Traefik will force the ACME certificate generation even if wildcard certificate exists.

The code should look like this :

```go
func (a *ACME) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
	domain := types.CanonicalDomain(clientHello.ServerName)
	account := a.store.Get().(*Account)

      // Check if we have a wildcard cert into TLSConfig that could be used
     // Only if the domain has not to filtered by wildcard certificates
	if a.isDomainFiltered(domain) {
		log.Debugf("Checking wildcard certificate deactived for domain %v", domain)
	} else {
		log.Debugf("Checking wildcard certificate matching for %v", domain)
		for k := range a.TLSConfig.NameToCertificate {
			selector := "^" + strings.Replace(k, "*.", "[^\\.]*\\.?", -1) + "$"
			match, _ := regexp.MatchString(selector, domain)
			if match {
				log.Debugf("Found a wildcard certificate to use as fallback: %v", k)
				return a.TLSConfig.NameToCertificate[k], nil
			}
		}
	}
    
	// Check for existing challenge cert
	log.Debugf("Checking ACME challenge for %s", domain)
	if challengeCert, ok := a.challengeProvider.getCertificate(domain); ok {
		log.Debugf("Returning ACME challenge cert for %s", domain)
		return challengeCert, nil
	}

	// Check for existing ACME domain cert
	log.Debugf("Checking ACME domain cert for %v", domain)
	if domainCert, ok := account.DomainsCertificate.getCertificateForDomain(domain); ok {
		log.Debugf("Returning ACME domain cert for %s", domain)
		return domainCert.tlsCert, nil
	}

	if a.OnDemand {
		if a.checkOnDemandDomain != nil && !a.checkOnDemandDomain(domain) {
			return nil, nil
		}
		cert, err := a.loadCertificateOnDemand(clientHello)
		if cert != nil && err == nil {
			return cert, nil
		}
		if err != nil {
			log.Errorf("On demand certificate retrieval failed for %v due to %v", domain, err)
		} else {
			log.Warnf("On demand certificate for %v didn't return anything.", domain)
		}
	}

	log.Debugf("No certificate found to return for %s", domain)
	return nil, nil
}
``` 
@CyrilPeponnet, what is your mind about this new suggestion?
cc/ @dtomcej